Memcached Backend

marczych opened the pull request:
This adds the `Memcached` backend which uses the more modern
[Memcached extension](http://php.net/manual/en/book.memcached.php).

I had to modify a few of the tests to get them to pass on HHVM. I filed an
issue for a few of them at facebook/hhvm#5121 so we'll see what happens
with that. This library can only be as good as the underlying extensions
so I think that the minor test modifications are fine.

danielbeardsley commented:
So sad that I somehow missed this for nearly a month!

This is great. CR :+1: 

Was there any good news in the benchmarks?

marczych commented:
> Was there any good news in the benchmarks?

`Memcached` appears to be slightly slower than `Memcache` on my local machine:

```sh
[mzych@Marcs-Air] - [~/Documents/Matryoshka] - [Wed May 06, 06:43:13]
[N]> php benchmarks/MatryoshkaBenchmark.php --backend=^Memcache
                                     Memcache           Memcached
                     Hits            0.036098            0.050040
 GetAndSetMultipleNone100            0.037878            0.055939
                 HalfHits            0.037915            0.051456
 GetAndSetMultipleHalf100            0.038942            0.054352
GetAndSetMultipleHalf1000            0.040060            0.055093
  GetAndSetMultipleAll100            0.040964            0.051326
       GetIncrementalKeys            0.041546            0.051406
       SetIncrementalKeys            0.041806            0.051972
[mzych@Marcs-Air] - [~/Documents/Matryoshka] - [Wed May 06, 06:43:24]
[I]> git stash pop
On branch backend--delete-multiple
Your branch is up-to-date with 'origin/backend--delete-multiple'.
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   library/iFixit/Matryoshka/Memcache.php
        modified:   library/iFixit/Matryoshka/Memcached.php

no changes added to commit (use "git add" and/or "git commit -a")
Dropped refs/stash@{0} (997624bdaee58b37d5af4c33f0f7544e0bee386b)
[mzych@Marcs-Air] - [~/Documents/Matryoshka] - [Wed May 06, 06:43:28]
[I]> git diff
diff --git a/library/iFixit/Matryoshka/Memcache.php b/library/iFixit/Matryoshka/Memcache.php
index 7fc3765..0e76f31 100644
--- a/library/iFixit/Matryoshka/Memcache.php
+++ b/library/iFixit/Matryoshka/Memcache.php
@@ -20,6 +20,7 @@ class Memcache extends Backend {
     * truncated.
     */
    public static function create(\Memcache $memcache) {
+      return new self($memcache);
       return new KeyShorten(new self($memcache), self::MAX_KEY_LENGTH);
    }

diff --git a/library/iFixit/Matryoshka/Memcached.php b/library/iFixit/Matryoshka/Memcached.php
index ffeead0..dd5dcfe 100644
--- a/library/iFixit/Matryoshka/Memcached.php
+++ b/library/iFixit/Matryoshka/Memcached.php
@@ -18,6 +18,7 @@ class Memcached extends Backend {
     * truncated.
     */
    public static function create(\Memcached $memcached) {
+      return new self($memcached);
       return new KeyShorten(new self($memcached), self::MAX_KEY_LENGTH);
    }

[mzych@Marcs-Air] - [~/Documents/Matryoshka] - [Wed May 06, 06:43:33]
[N]> php benchmarks/MatryoshkaBenchmark.php --backend=^Memcache
                                     Memcache           Memcached
                     Hits            0.032301            0.048227
                 HalfHits            0.033496            0.048210
 GetAndSetMultipleHalf100            0.035483            0.048761
  GetAndSetMultipleAll100            0.037180            0.052338
       SetIncrementalKeys            0.037514            0.048163
GetAndSetMultipleHalf1000            0.037685            0.053955
 GetAndSetMultipleNone100            0.039475            0.049664
       GetIncrementalKeys            0.040368            0.049419
```
